### PR TITLE
Add shutdown strategy

### PR DESCRIPTION
**Pull request by rjharrison**
## Goal

We currently rely on register_shutdown_function() to defer calls to flush(). This potentially causes a [memory leak](https://github.com/bugsnag/bugsnag-symfony/issues/77) as references to the Client are held until the script exits. (This is only an issue when testing with many instances of Client objects). 

The call to register_shutdown_function is hardcoded in the constructor. This PR introduces a ShutdownStrategyInterface which replaces this hardcoding and allows for the injection of different shutdown strategies. 

There is no functionality change in this PR. The current register_shutdown_function() behaviour is now wrapped inside an object. In a future PR on the bugsnag-symfony project I will introduce a Symfony-specific shutdown strategy that will fix [issue-77](https://github.com/bugsnag/bugsnag-symfony/issues/77).

### Added

- New ShutdownStrategyInterface parameter to Client constructor (defaults to null)
- PhpShutdownStrategy (replicates current behaviour)

### Removed

- hardcoded call to register_shutdown_function

## Tests

Added unit tests. The "shutdownSpy" is a pretty cool way to test native functions (in this case the call to register_shutdown_function)

### Alternative Approaches

None.

### Outstanding Questions

I think we should also make the call to registerShutdownStrategy conditional on whether we are in batch mode or not (there is no need to defer calls to flush() if we are not "batching"). However at this stage I want to change one thing at a time, so will potentially do this as a separate piece of work.

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -107,10 +109,11 @@ public static function make($apiKey = null, $endpoint = null, $defaults = true)
      * @param \Bugsnag\Configuration                  $config
      * @param \Bugsnag\Request\ResolverInterface|null $resolver
      * @param \GuzzleHttp\ClientInterface|null        $guzzle
+     * @param ShutdownStrategyInterface               $shutdownStrategy
```
Comment (GrahamCampbell) [inline src/Client.php]: Missing `|null`.

Comment (GrahamCampbell) [reply src/Client.php]: Also, should use a FQCN out of convention.

File: src/Shutdown/ShutdownStrategyInterface.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * Interface ShutdownStrategyInterface.
+ */
+interface ShutdownStrategyInterface
+{
+    /**
+     * Register the shutdown behaviour.
+     *
+     * @param Client $client
+     */
```
Comment (GrahamCampbell) [inline src/Shutdown/ShutdownStrategyInterface.php]: Should use a FQCN. Also, missing a return void annotations to let people know that they are not expected to return anything, Vs undocumented.

File: src/Client.php (hunk 1 above)
Comment (rjharrison) [reply src/Client.php]: Fixed, thanks.

File: src/Shutdown/ShutdownStrategyInterface.php (hunk 2 above)
Comment (rjharrison) [reply src/Shutdown/ShutdownStrategyInterface.php]: Fixed, thanks.

File: src/Shutdown/ShutdownStrategyInterface.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Bugsnag\Shutdown;
+
+use Bugsnag\Client;
+
+/**
+ * Interface ShutdownStrategyInterface.
```
Comment (tomlongridge) [inline src/Shutdown/ShutdownStrategyInterface.php]: Worth expanding the docs here to describe the purpose of the interface and expected reasons for implementing it?
